File: RadarSDKTests/RadarHostFailoverTests.m
//
//  RadarHostFailoverTests.m
//  RadarSDKTests
//
//  Copyright © 2025 Radar Labs, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "RadarHostFailover.h"

@interface RadarHostFailoverTests : XCTestCase
@end

@implementation RadarHostFailoverTests

static NSString *const kPrimary = @"api.radar.io";
static NSString *const kFallback1 = @"api-fallback1.radar.io";
static NSString *const kFallback2 = @"api-fallback2.radar.io";

#pragma mark - Initial state

- (void)test_currentHost_returnsPrimaryInitially {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    XCTAssertEqualObjects([failover currentHost], kPrimary);
}

- (void)test_currentHost_singleHost {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary]];
    XCTAssertEqualObjects([failover currentHost], kPrimary);
}

#pragma mark - reportFailure

- (void)test_reportFailure_returnsTrueWhenAlternateAvailable {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    BOOL hasAlternate = [failover reportFailure];
    XCTAssertTrue(hasAlternate);
}

- (void)test_reportFailure_switchesToFallback {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1, kFallback2]];
    [failover reportFailure];
    XCTAssertEqualObjects([failover currentHost], kFallback1);
}

- (void)test_reportFailure_returnsFalseOnLastHost {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary]];
    BOOL hasAlternate = [failover reportFailure];
    XCTAssertFalse(hasAlternate);
}

- (void)test_reportFailure_advancesThroughAllHosts {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1, kFallback2]];

    BOOL first = [failover reportFailure];
    XCTAssertTrue(first);
    XCTAssertEqualObjects([failover currentHost], kFallback1);

    BOOL second = [failover reportFailure];
    XCTAssertTrue(second);
    XCTAssertEqualObjects([failover currentHost], kFallback2);

    BOOL third = [failover reportFailure];
    XCTAssertFalse(third);
}

#pragma mark - reportSuccess

- (void)test_reportSuccess_normalModeDoesNotChangeHost {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportSuccess];
    XCTAssertEqualObjects([failover currentHost], kPrimary);
}

- (void)test_reportSuccess_notProbingStaysOnFallback {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure];
    // Within backoff so not probing; reportSuccess should not reset to primary
    [failover reportSuccess];
    XCTAssertEqualObjects([failover currentHost], kFallback1);
}

#pragma mark - Probing behavior

- (void)test_currentHost_probesPrimaryAfterBackoff {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure];

    // Simulate backoff elapsed
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-60] forKey:@"lastFailureTime"];

    XCTAssertEqualObjects([failover currentHost], kPrimary);
}

- (void)test_reportSuccess_afterProbeResetsToPrimary {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure];

    // Simulate backoff elapsed and trigger probe
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-60] forKey:@"lastFailureTime"];
    NSString *host = [failover currentHost];
    XCTAssertEqualObjects(host, kPrimary);

    // Probe succeeds
    [failover reportSuccess];
    XCTAssertEqualObjects([failover currentHost], kPrimary);
}

- (void)test_reportFailure_probeFailedDoublesBackoff {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure];

    // Simulate backoff elapsed to trigger probe
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-60] forKey:@"lastFailureTime"];
    [failover currentHost]; // sets isProbingPrimary = YES

    // Probe fails
    BOOL hasAlternate = [failover reportFailure];
    XCTAssertTrue(hasAlternate);

    // Backoff should have doubled from 30 to 60
    NSTimeInterval currentBackoff = [[failover valueForKey:@"currentBackoff"] doubleValue];
    XCTAssertEqual(currentBackoff, 60.0);
}

- (void)test_backoffCapsAt300Seconds {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure];

    // Repeatedly probe and fail: 30 -> 60 -> 120 -> 240 -> 300 (cap)
    for (int i = 0; i < 10; i++) {
        [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-600] forKey:@"lastFailureTime"];
        [failover currentHost]; // trigger probe
        [failover reportFailure]; // probe fails, doubles backoff
    }

    NSTimeInterval currentBackoff = [[failover valueForKey:@"currentBackoff"] doubleValue];
    XCTAssertEqual(currentBackoff, 300.0);
}

- (void)test_reportSuccess_resetsBackoffToInitial {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure];

    // Drive backoff up
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-60] forKey:@"lastFailureTime"];
    [failover currentHost];
    [failover reportFailure]; // backoff now 60

    // Probe again and succeed
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-120] forKey:@"lastFailureTime"];
    [failover currentHost];
    [failover reportSuccess];

    NSTimeInterval currentBackoff = [[failover valueForKey:@"currentBackoff"] doubleValue];
    XCTAssertEqual(currentBackoff, 30.0);
}

#pragma mark - Fallback during backoff

- (void)test_currentHost_returnsFallbackDuringBackoff {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure];

    // lastFailureTime is now, so backoff hasn't elapsed
    XCTAssertEqualObjects([failover currentHost], kFallback1);
    XCTAssertEqualObjects([failover currentHost], kFallback1);
}

#pragma mark - Full failover cycle

- (void)test_fullCycle_failoverProbeRecover {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1, kFallback2]];

    // 1. Start on primary
    XCTAssertEqualObjects([failover currentHost], kPrimary);

    // 2. Primary fails, move to fallback1
    [failover reportFailure];
    XCTAssertEqualObjects([failover currentHost], kFallback1);

    // 3. Backoff elapses, probe primary
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-60] forKey:@"lastFailureTime"];
    XCTAssertEqualObjects([failover currentHost], kPrimary);

    // 4. Probe fails, stay on fallback1 with doubled backoff
    [failover reportFailure];
    XCTAssertEqualObjects([failover currentHost], kFallback1);

    // 5. Backoff elapses again, probe primary
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-120] forKey:@"lastFailureTime"];
    XCTAssertEqualObjects([failover currentHost], kPrimary);

    // 6. Probe succeeds, back to primary
    [failover reportSuccess];
    XCTAssertEqualObjects([failover currentHost], kPrimary);
}

- (void)test_failoverToSecondFallbackThenRecover {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1, kFallback2]];

    // Fail through to fallback2
    [failover reportFailure]; // -> fallback1
    [failover reportFailure]; // -> fallback2
    XCTAssertEqualObjects([failover currentHost], kFallback2);

    // Backoff elapses, probe primary, succeeds
    [failover setValue:[NSDate dateWithTimeIntervalSinceNow:-60] forKey:@"lastFailureTime"];
    XCTAssertEqualObjects([failover currentHost], kPrimary);
    [failover reportSuccess];

    // Fully recovered
    XCTAssertEqualObjects([failover currentHost], kPrimary);
}

- (void)test_reportFailure_onLastHostReturnsFalseRepeatedly {
    RadarHostFailover *failover = [[RadarHostFailover alloc] initWithHosts:@[kPrimary, kFallback1]];
    [failover reportFailure]; // -> fallback1
    BOOL result1 = [failover reportFailure]; // already on last host
    BOOL result2 = [failover reportFailure]; // still on last host
    XCTAssertFalse(result1);
    XCTAssertFalse(result2);
}

@end

File: RadarSDKTests/RadarHostFailoverTests.swift
//
//  RadarHostFailoverTests.swift
//  RadarSDKTests
//
//  Copyright © 2025 Radar Labs, Inc. All rights reserved.
//

import Testing
@testable import RadarSDK

@Suite(.serialized)
struct RadarHostFailoverTests {

    let primary = "api.radar.io"
    let fallback1 = "api-fallback1.radar.io"
    let fallback2 = "api-fallback2.radar.io"

    // MARK: - Initial state

    @Test func currentHost_returnsPrimaryInitially() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        #expect(failover.currentHost == primary)
    }

    @Test func currentHost_singleHost() {
        let failover = RadarHostFailover(hosts: [primary])
        #expect(failover.currentHost == primary)
    }

    // MARK: - reportFailure

    @Test func reportFailure_returnsTrueWhenAlternateAvailable() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        #expect(failover.reportFailure() == true)
    }

    @Test func reportFailure_switchesToFallback() {
        let failover = RadarHostFailover(hosts: [primary, fallback1, fallback2])
        failover.reportFailure()
        #expect(failover.currentHost == fallback1)
    }

    @Test func reportFailure_returnsFalseOnLastHost() {
        let failover = RadarHostFailover(hosts: [primary])
        #expect(failover.reportFailure() == false)
    }

    @Test func reportFailure_advancesThroughAllHosts() {
        let failover = RadarHostFailover(hosts: [primary, fallback1, fallback2])

        #expect(failover.reportFailure() == true)
        #expect(failover.currentHost == fallback1)

        #expect(failover.reportFailure() == true)
        #expect(failover.currentHost == fallback2)

        #expect(failover.reportFailure() == false)
    }

    // MARK: - reportSuccess

    @Test func reportSuccess_normalModeDoesNotChangeHost() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportSuccess()
        #expect(failover.currentHost == primary)
    }

    @Test func reportSuccess_notProbingStaysOnFallback() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure()
        // Within backoff so not probing; reportSuccess should not reset to primary
        failover.reportSuccess()
        #expect(failover.currentHost == fallback1)
    }

    // MARK: - Probing behavior

    @Test func currentHost_probesPrimaryAfterBackoff() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure()

        // Simulate backoff elapsed
        failover.lastFailureTime = Date(timeIntervalSinceNow: -60)

        #expect(failover.currentHost == primary)
    }

    @Test func reportSuccess_afterProbeResetsToPrimary() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure()

        // Simulate backoff elapsed and trigger probe
        failover.lastFailureTime = Date(timeIntervalSinceNow: -60)
        #expect(failover.currentHost == primary)

        // Probe succeeds
        failover.reportSuccess()
        #expect(failover.currentHost == primary)
    }

    @Test func reportFailure_probeFailedDoublesBackoff() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure()

        // Simulate backoff elapsed to trigger probe
        failover.lastFailureTime = Date(timeIntervalSinceNow: -60)
        _ = failover.currentHost // sets isProbingPrimary = true

        // Probe fails
        #expect(failover.reportFailure() == true)

        // Backoff should have doubled from 30 to 60
        #expect(failover.currentBackoff == 60.0)
    }

    @Test func backoffCapsAt300Seconds() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure()

        // Repeatedly probe and fail: 30 -> 60 -> 120 -> 240 -> 300 (cap)
        for _ in 0..<10 {
            failover.lastFailureTime = Date(timeIntervalSinceNow: -600)
            _ = failover.currentHost // trigger probe
            failover.reportFailure() // probe fails, doubles backoff
        }

        #expect(failover.currentBackoff == 300.0)
    }

    @Test func reportSuccess_resetsBackoffToInitial() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure()

        // Drive backoff up
        failover.lastFailureTime = Date(timeIntervalSinceNow: -60)
        _ = failover.currentHost
        failover.reportFailure() // backoff now 60

        // Probe again and succeed
        failover.lastFailureTime = Date(timeIntervalSinceNow: -120)
        _ = failover.currentHost
        failover.reportSuccess()

        #expect(failover.currentBackoff == 30.0)
    }

    // MARK: - Fallback during backoff

    @Test func currentHost_returnsFallbackDuringBackoff() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure()

        // lastFailureTime is now, so backoff hasn't elapsed
        #expect(failover.currentHost == fallback1)
        #expect(failover.currentHost == fallback1)
    }

    // MARK: - Full failover cycle

    @Test func fullCycle_failoverProbeRecover() {
        let failover = RadarHostFailover(hosts: [primary, fallback1, fallback2])

        // 1. Start on primary
        #expect(failover.currentHost == primary)

        // 2. Primary fails, move to fallback1
        failover.reportFailure()
        #expect(failover.currentHost == fallback1)

        // 3. Backoff elapses, probe primary
        failover.lastFailureTime = Date(timeIntervalSinceNow: -60)
        #expect(failover.currentHost == primary)

        // 4. Probe fails, stay on fallback1 with doubled backoff
        failover.reportFailure()
        #expect(failover.currentHost == fallback1)

        // 5. Backoff elapses again, probe primary
        failover.lastFailureTime = Date(timeIntervalSinceNow: -120)
        #expect(failover.currentHost == primary)

        // 6. Probe succeeds, back to primary
        failover.reportSuccess()
        #expect(failover.currentHost == primary)
    }

    @Test func failoverToSecondFallbackThenRecover() {
        let failover = RadarHostFailover(hosts: [primary, fallback1, fallback2])

        // Fail through to fallback2
        failover.reportFailure() // -> fallback1
        failover.reportFailure() // -> fallback2
        #expect(failover.currentHost == fallback2)

        // Backoff elapses, probe primary, succeeds
        failover.lastFailureTime = Date(timeIntervalSinceNow: -60)
        #expect(failover.currentHost == primary)
        failover.reportSuccess()

        // Fully recovered
        #expect(failover.currentHost == primary)
    }

    @Test func reportFailure_onLastHostReturnsFalseRepeatedly() {
        let failover = RadarHostFailover(hosts: [primary, fallback1])
        failover.reportFailure() // -> fallback1
        #expect(failover.reportFailure() == false) // already on last host
        #expect(failover.reportFailure() == false) // still on last host
    }
}
